sdk 2.0.3版本下必现panic异常

启动监听服务状态

panic: reflect: call of reflect. Value. Type on zero Value (recovered]
panic: reflect: call of reflect.Value.Type on zero Value
goroutine 1287 [running]:
encoding/json. (*encodeState) . marshal. func1 (0xc000659a08)
/usr/local/go/src/encoding/json/encode.go:328+0x8d
panic(Oxfaece0, Oxc0002ea9a8)
/usr/local/go/src/runtime/panic.go:965+0x1b9
reflect.Value.Type(0x0, 0x0, 0x0, Oxc000f9bOel, 0x0)
/usr/local/go/src/reflect/value.go:1908+0x189
1
encoding/json.stringEncoder(0xc00122a200,0x0,0x0,0x0,0xc000f90100)
2
/usr/local/go/src/encoding/json/encode.go:620+0x46
13
encoding/json.mapEncoder-encode(0x123c888,0xc00122a200,0xfbd880,0xc0015daf68,0x195,0x100)
14
/usr/local/go/src/encoding/json/encode.go:813+0x366
15
encoding/json.structEncoder-encode(0xc000517b00, Oxd, 0x10, Oxc0001cda70, Oxc00122a200, 0x10c6860, Oxc0015daf10, 0x199, 0x100)
16
/usr/local/go/src/encoding/json/encode.go:761+0x2ab
17
encoding/json.arrayEncoder.encode(0xc0001cdaa®, Oxc00122a200, Oxf566a0, Oxc00045a858, 0x97, 0xb60100)
18
/usr/local/go/src/encoding/json/encode.go:915+0xd6
19
encoding/json.sliceEncoder-encode(0xc0003bccf0,0xc00122a200,Oxf566a0,0xc00045a858,0x97,0x100)
20
21
/usr/local/go/src/encoding/json/encode.go:888+0xfe
encoding/json.typeEncoder.func1(0xc00122a200,0xf566a0,0xc00045a858,0x97,0x100)
22
/usr/local/go/src/encoding/json/encode.go:397+0x79
23
encoding/json.structEncoder.encode(0xc00038a900,0xa,0x10,0xc0002f2030,0xc00122a200,0x10b43e0,0xc000452850,0x99,0x10b0100)
24
/usr/local/go/src/encoding/json/encode.go:761+0x2ab
25
encoding/json. (*encodeState). reflectValue(0xc00122a200, 0x10b43e0, Oxc00045a850, 0x99, 0x000650100)
26
/usr/local/go/src/encoding/json/encode.go:360+0x82
27
encoding/json. (*encodeState) .marshal(0xc00122a200, 0x10b43e0, Oxc00045a850, 0x100, 0x0, 0x0)
28
/usr/local/go/src/encoding/json/encode.go:332+0xf9
29
encoding/json.Marshal(0x10b43e0, Oxc00045a850, 0x10b43e0, Oxc00045a850, 0x0, 0x0, 0xc0003360c0)
30
/usr/local/go/src/encoding/json/encode.go:161+0x52
31
github.com/nacos-group/nacos-sdk-go/v2/clients/cache.WriteServicesToFile(0x2710, Oxc0015da880, Ox1a, 0x1c, 0x0, 0x0, 0x1814d062de3, 0x0, 0x0, 0xc000334520,
...)
32
/opt/devel/vendor/github.com/nacos-group/nacos-sdk-go/v2/clients/cache/disk_cache.go:39+0xa8
33
34
github.com/nacos-group/nacos-sdk-go/v2/clients/naming_client/naming_cache. (#ServiceInfoHolder).ProcessService(0xc00011ad20,0xc0007h0888)
/opt/devel/vendor/github.com/nacos-group/nacos-sdk-go/v2/clients/naming_client/naming_cache/service_info_holder.go:97+0x7a5
35
36
github.com/nacos-group/nacos-sdk-go/v2/common/remote/rpc. (*NamingPushRequestHandler)•RequestReply(0xc000011698,0x1357038,0xc0007b0880,0xc00050aa00,0xc0007be
37
/opt/devel/vendor/github.com/nacos-group/nacos-sdk-go/v2/common/remote/rpc/server_request_handler.go:93+0x57
github.com/nacos-group/nacos-sdk-go/v2/common/remote/rpc.(#GrpcClient).handleServerRequest(0xc0000115f8,0xc000a66e10,0xc00031edb0)
38
/opt/devel/vendor/github.com/nacos-group/nacos-sdk-go/v2/common/remote/rpc/grpc_client.go:246+0x1aa
39
github.com/nacos-group/nacos-sdk-go/v2/common/remote/rpc. (#GrpcClient).bindBiRequestStream.func1(0x135da40,0xc000147780,0xc0000115f8,0xc00031edb0)
40
/.opt/devel/vendor/github.com/nacos-group/nacos-sdk-go/v2/common/remote/rpc/grpc_client.go:193+0x4d9
41
created by github.com/nacos-group/nacos-sdk-go/v2/common/remote/rpc. (*GrpcClient).bindBiRequestStream
42
/opt/devel/vendor/github.com/nacos-group/nacos-sdk-go/v2/common/remote/rpc/grpc_client.go:168+0x5d
43

--- a/clients/naming_client/naming_cache/service_info_holder.go
+++ b/clients/naming_client/naming_cache/service_info_holder.go
@@ -92,7 +92,7 @@ func (s *ServiceInfoHolder) ProcessService(service *model.Service) {
 
 	s.UpdateTimeMap.Set(cacheKey, uint64(util.CurrentMillis()))
 	s.ServiceInfoMap.Set(cacheKey, *service)
-	if !ok || checkInstanceChanged(oldDomain, service) {
+	if !ok || checkInstanceChanged(oldDomain, *service) {
 		logger.Infof("service key:%s was updated to:%s", cacheKey, util.ToJsonString(service))
 		cache.WriteServicesToFile(service, cacheKey, s.cacheDir)
 		s.subCallback.ServiceChanged(cacheKey, service)
@@ -127,16 +127,16 @@ func (s *ServiceInfoHolder) IsSubscribed(serviceName, clusters string) bool {
 	return s.subCallback.IsSubscribed(serviceName, clusters)
 }
 
-func checkInstanceChanged(oldDomain interface{}, service *model.Service) bool {
+func checkInstanceChanged(oldDomain interface{}, service model.Service) bool {
 	if oldDomain == nil {
 		return true
 	}
 	oldService := oldDomain.(model.Service)
-	return isServiceInstanceChanged(&oldService, service)
+	return isServiceInstanceChanged(oldService, service)
 }
 
 // return true when service instance changed ,otherwise return false.
-func isServiceInstanceChanged(oldService, newService *model.Service) bool {
+func isServiceInstanceChanged(oldService, newService model.Service) bool {
 	oldHostsLen := len(oldService.Hosts)
 	newHostsLen := len(newService.Hosts)
 	if oldHostsLen != newHostsLen {

--- a/clients/naming_client/naming_cache/service_info_holder_test.go
+++ b/clients/naming_client/naming_cache/service_info_holder_test.go
@@ -101,7 +101,7 @@ func TestServiceInfoHolder_isServiceInstanceChanged(t *testing.T) {
 
 func TestHostReactor_isServiceInstanceChangedWithUnOrdered(t *testing.T) {
 	rand.Seed(time.Now().Unix())
-	serviceA := &model.Service{
+	serviceA := model.Service{
 		LastRefTime: 1001,
 		Hosts: []model.Instance{
 			{
@@ -119,7 +119,7 @@ func TestHostReactor_isServiceInstanceChangedWithUnOrdered(t *testing.T) {
 		},
 	}
 
-	serviceB := &model.Service{
+	serviceB := model.Service{
 		LastRefTime: 1001,
 		Hosts: []model.Instance{
 			{

--- a/clients/naming_client/naming_cache/service_info_holder_test.go
+++ b/clients/naming_client/naming_cache/service_info_holder_test.go
@@ -30,7 +30,7 @@ func TestServiceInfoHolder_isServiceInstanceChanged(t *testing.T) {
 	rand.Seed(time.Now().Unix())
 	defaultIp := createRandomIp()
 	defaultPort := creatRandomPort()
-	serviceA := &model.Service{
+	serviceA := model.Service{
 		LastRefTime: 1000,
 		Hosts: []model.Instance{
 			{
@@ -47,7 +47,7 @@ func TestServiceInfoHolder_isServiceInstanceChanged(t *testing.T) {
 			},
 		},
 	}
-	serviceB := &model.Service{
+	serviceB := model.Service{
 		LastRefTime: 1001,
 		Hosts: []model.Instance{
 			{
@@ -65,7 +65,7 @@ func TestServiceInfoHolder_isServiceInstanceChanged(t *testing.T) {
 		},
 	}
 	ip := createRandomIp()
-	serviceC := &model.Service{
+	serviceC := model.Service{
 		LastRefTime: 1001,
 		Hosts: []model.Instance{
 			{

